### PR TITLE
Simplify the `tcx.alloc_map` API

### PR DESCRIPTION
**Pull request by oli-obk**
This PR changes all functions that require manually locking the `alloc_map` to functions on `TyCtxt` that lock the map internally. In the same step we make the `TyCtxt::alloc_map` field private.

r? @RalfJung

### CODE REVIEW

File: src/librustc_middle/mir/interpret/mod.rs [hunk 1]
```diff
@@ -398,12 +396,6 @@ impl<'tcx> AllocMap<'tcx> {
     pub fn new() -> Self {
         AllocMap { alloc_map: Default::default(), dedup: Default::default(), next_id: AllocId(0) }
     }
-
-    /// Obtains a new allocation ID that can be referenced but does not
-    /// yet have an allocation backing it.
-    ///
-    /// Make sure to call `set_alloc_id_memory` or `set_alloc_id_same_memory` before returning such
-    /// an `AllocId` from a query.
     pub fn reserve(&mut self) -> AllocId {
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/mod.rs:399]: Can this become private now? Otherwise, why did you remove the doc-comment?

File: src/librustc_middle/mir/interpret/mod.rs [hunk 2]
```diff
@@ -454,8 +458,9 @@ impl<'tcx> AllocMap<'tcx> {
         });
         if is_generic {
             // Get a fresh ID.
-            let id = self.reserve();
-            self.alloc_map.insert(id, GlobalAlloc::Function(instance));
+            let mut alloc_map = self.alloc_map.lock();
+            let id = alloc_map.reserve();
+            alloc_map.alloc_map.insert(id, GlobalAlloc::Function(instance));
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/mod.rs:463]: `alloc_map.alloc_map` looks odd.

File: src/librustc_middle/mir/interpret/mod.rs [hunk 3]
```diff
@@ -480,38 +485,38 @@ impl<'tcx> AllocMap<'tcx> {
...
     pub fn unwrap_memory(&self, id: AllocId) -> &'tcx Allocation {
-        match self.get(id) {
+        match self.get_global_alloc(id) {
             Some(GlobalAlloc::Memory(mem)) => mem,
             _ => bug!("expected allocation ID {} to point to memory", id),
         }
     }
 
     /// Panics if the `AllocId` does not refer to a function
     pub fn unwrap_fn(&self, id: AllocId) -> Instance<'tcx> {
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/mod.rs:501]: As a tcx method, this name is not very descriptive. Same for `unwrap_memory`.

What about `unwrap_global_alloc_memory` or so? Something involving `global_alloc`, or at least `global`.

File: src/librustc_middle/mir/interpret/mod.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc_middle/mir/interpret/mod.rs:399]: oh right. can be private, I moved the doc comment to a function on `TyCtxt`. This function is still on `AllocMap`.

File: src/librustc_middle/mir/interpret/mod.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_middle/mir/interpret/mod.rs:463]: it does, but it's only inside the `AllocMap` impls

File: src/librustc_middle/mir/interpret/mod.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc_middle/mir/interpret/mod.rs:501]: I assumed that its argument made that obvious... I was honestly playing with the thought of allowing indexing `TyCtxt` with `AllocId` to get `GlobalAlloc`.

File: src/librustc_middle/mir/interpret/mod.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc_middle/mir/interpret/mod.rs:463]: Where does it come from though?

File: src/librustc_middle/mir/interpret/mod.rs (hunk 3 above)
Comment (RalfJung) [reply src/librustc_middle/mir/interpret/mod.rs:501]: The argument type might not be obvious at the call site though.

Comment (RalfJung) [reply src/librustc_middle/mir/interpret/mod.rs:501]: Alternatively, what about `get_global_alloc(...).unwrap_[fn|memory]()`?

File: src/librustc_middle/mir/interpret/mod.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_middle/mir/interpret/mod.rs:463]: There is `TyCtxt::alloc_map` of `AllocMap` type, and then there's `AllocMap::alloc_map` of `FxHashMap` type.

File: src/librustc_middle/mir/interpret/mod.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc_middle/mir/interpret/mod.rs:501]: oh heh, that's nice

File: src/librustc_middle/mir/interpret/mod.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc_middle/mir/interpret/mod.rs:463]: Hm... there might be better names but it's probably not worth renaming.^^

File: src/librustc_middle/ty/print/pretty.rs [hunk 4]
```diff
@@ -1022,7 +1024,7 @@ pub trait PrettyPrinter<'tcx>:
                 )?;
             }
             (Scalar::Ptr(ptr), ty::FnPtr(_)) => {
-                let instance = self.tcx().unwrap_fn(ptr.alloc_id);
+                let instance = self.tcx().get_global_alloc(ptr.alloc_id).unwrap().unwrap_fn();
```
Comment (RalfJung) [inline src/librustc_middle/ty/print/pretty.rs:1027]: Ah I forgot that we now have double-unwraps... hm. Not sure if it is worth trying to do something about them?

Comment (oli-obk) [reply src/librustc_middle/ty/print/pretty.rs:1027]: `self.tcx()[ptr.alloc_id]` would solve it ;)

Comment (RalfJung) [reply src/librustc_middle/ty/print/pretty.rs:1027]: A method `global_alloc(alloc_id)` that panics on not-found would, too. ;)

Comment (oli-obk) [reply src/librustc_middle/ty/print/pretty.rs]: done

File: src/librustc_middle/mir/interpret/mod.rs [hunk 5]
```diff
@@ -482,38 +509,31 @@ impl<'tcx> AllocMap<'tcx> {
...
     }
 
-    /// Panics if the `AllocId` does not refer to a function
-    pub fn unwrap_fn(&self, id: AllocId) -> Instance<'tcx> {
-        match self.get(id) {
-            Some(GlobalAlloc::Function(instance)) => instance,
-            _ => bug!("expected allocation ID {} to point to a function", id),
+    #[inline]
+    #[track_caller]
+    pub fn global_alloc(&self, id: AllocId) -> GlobalAlloc<'tcx> {
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/mod.rs:518]: Please add a doc comment.

File: src/librustc_middle/mir/mod.rs [hunk 6]
```diff
@@ -2410,13 +2410,9 @@ pub struct Constant<'tcx> {
 impl Constant<'tcx> {
     pub fn check_static_ptr(&self, tcx: TyCtxt<'_>) -> Option<DefId> {
         match self.literal.val.try_to_scalar() {
-            Some(Scalar::Ptr(ptr)) => match tcx.alloc_map.lock().get(ptr.alloc_id) {
-                Some(GlobalAlloc::Static(def_id)) => Some(def_id),
-                Some(_) => None,
-                None => {
-                    tcx.sess.delay_span_bug(DUMMY_SP, "MIR cannot contain dangling const pointers");
```
Comment (RalfJung) [inline src/librustc_middle/mir/mod.rs:2417]: This turns a delay_bug into a `bug!` -- let's see if we can get away with that. ;)

File: src/librustc_middle/mir/interpret/mod.rs [hunk 7]
```diff
@@ -381,7 +379,29 @@ pub enum GlobalAlloc<'tcx> {
...
+
+    /// Panics if the `GlobalAlloc` is not `GlobalAlloc::Function`
+    #[track_caller]
+    #[inline]
+    pub fn unwrap_fn(&self) -> Instance<'tcx> {
+        match *self {
+            GlobalAlloc::Function(instance) => instance,
+            _ => bug!("expected function, got {:?}", self),
+        }
+    }
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/mod.rs:401]: For symmetry, should we have `unwrap_static`?

Comment (oli-obk) [reply src/librustc_middle/mir/interpret/mod.rs:401]: I considered it, but found no possible use sites, thus left it out.

Comment (RalfJung) [reply src/librustc_middle/mir/interpret/mod.rs:401]: Fair enough.
